### ISSUE TITLE
Optional

### ISSUE THREAD

**Issue by Masynchin** [yegor256/takes#1143]
Hello, @yegor256!

Why don't this project use [`Optional`](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html), but rather implements its own `Opt`? I think that `Optional` can safely replace `Opt` and also bring `map` method, which can be used to get rid of `if opt.has() { /* do something with opt.get() */ } else { return Opt.Empty() }`.

For example, this method

https://github.com/yegor256/takes/blob/91d020c32348944f146c83957c365a82e6f1c433/src/main/java/org/takes/facets/auth/PsBasic.java#L238-L261

Can be rewrite like this:

~~~java
@Override
public Opt<Identity> enter(final String user, final String pwd) {
    return this.urn(user, pwd).map(urn -> {
        try {
            return new Identity.Simple(
                URLDecoder.decode(
                    urn, PsBasic.Default.ENCODING
                )
            );
        } catch (final UnsupportedEncodingException ex) {
            throw new IllegalStateException(
                String.format("Failed to decode URN '%s'", urn),
                ex
            );
        }
    });
}
~~~

It uses two less variables and one less `if-else` statement. I think this code is more readable and more maintainable.

**Comment by yegor256** [comment]
@Masynchin indeed, you are right. I can't remember why `Opt` was introduced. Maybe because we developed Takes for Java 6, where there was no `Option`. If you are interested, please, submit a pull request and let's get rid of `Opt`.

**Comment by yegor256** [comment]
@Masynchin you are planning to submit four pull requests?

**Comment by Masynchin** [comment]
> @Masynchin you are planning to submit four pull requests?

No, just one PR - "Replace Opt with Optional". This is just a step-by-step plan. Does this plan have any issues? If it seems OK, I can start implementing it.

**Comment by Masynchin** [comment]
It turned out I can't just create `OptionalEnvelope` as it `extends Optional implements Opt`. This refactoring will take much longer.

**Comment by yegor256** [comment]
`Opt` is an interface with two implementations (`Opt.Single`, `Opt.Empty`), so it composes and can be decorated. `java.util.Optional` is a final class reached through static factory methods — precisely what this project avoids, which is why it is not used anywhere in Takes.

The `map()` you propose would take a lambda, turning the object into a function holder. The same refactoring is done here by adding another `Opt` decorator instead.

Closing as answered — the choice is deliberate.
